platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +84,159 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+  }
+
+  void validateRateResponse(GetRateRequest request, GetRateResponse getRateResponse)
+      throws ServerErrorException {
+    AssetInfo sellAsset = assetService.getAssetByName(request.getSellAsset());
+    AssetInfo buyAsset = assetService.getAssetByName(request.getBuyAsset());
+
     GetRateResponse.Rate rate = getRateResponse.getRate();
     if (rate == null || rate.getPrice() == null) {
-      Log.error("missing 'price' in the GET /rate response");
-      throw new ServerErrorException("internal server error");
+      logErrorAndThrow("missing 'price' in the GET /rate response", ServerErrorException.class);
     }
 
     if (request.getType() != GetRateRequest.Type.INDICATIVE) {
-      if (rate.getFee() == null) {
-        Log.error("'fee' is missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+      if (Objects.requireNonNull(rate).getFee() == null) {
+        logErrorAndThrow("'fee' is missing in the GET /rate response", ServerErrorException.class);
       }
     }
 
     if (request.getType() == GetRateRequest.Type.FIRM) {
       if (rate.getId() == null || rate.getExpiresAt() == null) {
-        Log.error("'id' and/or 'expires_at' are missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+        logErrorAndThrow(
+            "'id' and/or 'expires_at' are missing in the GET /rate response",
+            ServerErrorException.class);
+      }
+    }
+
+    // sell_amount is present and positive number
+    if (!isPositiveNumber(Objects.requireNonNull(rate).getSellAmount())) {
+      logErrorAndThrow(
+          "'sell_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // sell_amount has a proper number of significant decimals
+    if (!NumberHelper.hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getSellAmount(), 0, sellAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          "'sell_amount' has incorrect number of significant decimals in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount is present and positive number
+    if (!isPositiveNumber(rate.getBuyAmount())) {
+      logErrorAndThrow(
+          "'buy_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount has a proper number of significant decimals
+    if (!NumberHelper.hasProperSignificantDecimals(
+        rate.getBuyAmount(), 0, buyAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          "'buy_amount' has incorrect number of significant decimals in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is preset, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
```

Comment:
It would be nice if this was a precondition for constructing a `FeeDetails` object.

Reply:
I mean including this validation in the `FeeDetails` constructor so it's impossible to initialize an instance with negative fees.

Reply:
ah, I see. 

The `FeeDetails` is in the api-schema sub-project. We did not put much logic in that specific sub-project. 

If we decide it is good to introduce that pattern in the code, I may create a ticket to introduce this and do a project wide review.